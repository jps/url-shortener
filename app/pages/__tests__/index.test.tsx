import { render, screen } from "@testing-library/react";
import React from "react";
import Home, { getServerSideProps } from "..";

let testUrls = ["https://jspenc.com/abcdefg", "https://jspenc.com/123456"];

const mockUseFetchUrlsQuery = jest.fn(() => ({
  data: { urls: [] as Array<string> },
}));

const mockUsePostUrlMutation = jest.fn(() => {
  return { mutate: jest.fn() };
});

const mockUseSavedUrl = jest.fn(() => ({
  isFetched: false,
}));

jest.mock("../../hooks", () => ({
  useFetchUrlsQuery: () => mockUseFetchUrlsQuery(),
  usePostUrlMutation: () => mockUsePostUrlMutation(),
  useSavedUrl: () => mockUseSavedUrl(),
}));

jest.mock("../../clients/url-client", () => ({
  fetchRecentUrlsInternal: async () => ({
    urls: testUrls,
  }),
}));

describe("index should", () => {
  describe("should render as expected", () => {
    const RendersAsExpected = () => {
      const { asFragment } = render(
        <Home
          recentUrls={{
            urls: [],
            status: "Success",
          }}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    };

    it("when no recent urls", () => {
      RendersAsExpected();
    });

    it("when recent urls", () => {
      mockUseFetchUrlsQuery.mockImplementationOnce(() => ({
        data: {
          urls: testUrls,
        },
      }));
      RendersAsExpected();
      screen.getByText("https://jspenc.com/abcdefg");
    });

    it("when success message", () => {
      mockUseSavedUrl.mockImplementationOnce(() => ({
        isFetched: true,
        data: {
          shortenedUrl: "https://jspenc.com/87654321",
          status: "Success",
        },
      }));

      RendersAsExpected();
      screen.getByText("https://jspenc.com/87654321");
    });
  });

  it("should get data for SSR", async () => {
    const result = await getServerSideProps();

    expect(result).toEqual({ props: { recentUrls: { urls: testUrls } } });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
