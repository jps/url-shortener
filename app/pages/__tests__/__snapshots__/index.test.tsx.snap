// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should should render as expected when no recent urls 1`] = `
<DocumentFragment>
  <form
    class="shorten-form"
  >
    <h2
      class="shorten-form__title"
    >
      Shorten URL
    </h2>
    <label
      class="visually-hidden"
      for="url"
    >
      Url
    </label>
    <input
      class="shorten-form__input form-text-input"
      id="url"
      name="url"
      placeholder="Shorten your link e.g. https://example.com"
      required=""
    />
    <input
      class="shorten-form__save form-submit"
      type="submit"
      value="Shorten"
    />
  </form>
</DocumentFragment>
`;

exports[`index should should render as expected when recent urls 1`] = `
<DocumentFragment>
  <form
    class="shorten-form"
  >
    <h2
      class="shorten-form__title"
    >
      Shorten URL
    </h2>
    <label
      class="visually-hidden"
      for="url"
    >
      Url
    </label>
    <input
      class="shorten-form__input form-text-input"
      id="url"
      name="url"
      placeholder="Shorten your link e.g. https://example.com"
      required=""
    />
    <input
      class="shorten-form__save form-submit"
      type="submit"
      value="Shorten"
    />
  </form>
  <div
    class="recent-urls"
  >
    <h2>
      Recent URLs
    </h2>
    <ul
      class="recent-urls__list"
    >
      <li
        class="recent-urls__list-item"
      >
        <a
          href="https://jspenc.com/abcdefg"
        >
          https://jspenc.com/abcdefg
        </a>
      </li>
      <li
        class="recent-urls__list-item"
      >
        <a
          href="https://jspenc.com/123456"
        >
          https://jspenc.com/123456
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`index should should render as expected when success message 1`] = `
<DocumentFragment>
  <form
    class="shorten-form"
  >
    <h2
      class="shorten-form__title"
    >
      Shorten URL
    </h2>
    <label
      class="visually-hidden"
      for="url"
    >
      Url
    </label>
    <input
      class="shorten-form__input form-text-input"
      id="url"
      name="url"
      placeholder="Shorten your link e.g. https://example.com"
      required=""
    />
    <input
      class="shorten-form__save form-submit"
      type="submit"
      value="Shorten"
    />
  </form>
  <div
    class="success-message"
  >
    <h2
      class="success-message__title"
    >
      Your URL has been shortened
    </h2>
    <p
      class="success-message__text"
    >
      <a
        href="https://jspenc.com/87654321"
      >
        https://jspenc.com/87654321
      </a>
    </p>
  </div>
</DocumentFragment>
`;
